lib/services/local/workload_identity.go

```diff
@@ -0,0 +1,115 @@
+)
+
+// WorkloadIdentityService exposes backend functionality for storing
+// WorkloadIdentity resources
+type WorkloadIdentityService struct {
+	service *generic.ServiceWrapper[*workloadidentityv1pb.WorkloadIdentity]
+}
+
+// NewWorkloadIdentityService creates a new WorkloadIdentityService
+func NewWorkloadIdentityService(b backend.Backend) (*WorkloadIdentityService, error) {
+	service, err := generic.NewServiceWrapper(
+		generic.ServiceWrapperConfig[*workloadidentityv1pb.WorkloadIdentity]{
+			Backend:       b,
+			ResourceKind:  types.KindWorkloadIdentity,
+			BackendPrefix: backend.NewKey(workloadIdentityPrefix),
+			MarshalFunc:   services.MarshalWorkloadIdentity,
+			UnmarshalFunc: services.UnmarshalWorkloadIdentity,
+			ValidateFunc:  services.ValidateWorkloadIdentity,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &WorkloadIdentityService{
+		service: service,
+	}, nil
+}
+
+// CreateWorkloadIdentity inserts a new WorkloadIdentity into the backend.
+func (b *WorkloadIdentityService) CreateWorkloadIdentity(
+	ctx context.Context, resource *workloadidentityv1pb.WorkloadIdentity,
+) (*workloadidentityv1pb.WorkloadIdentity, error) {
+	created, err := b.service.CreateResource(ctx, resource)
+	return created, trace.Wrap(err)
+}
+
+// GetWorkloadIdentity retrieves a specific WorkloadIdentity given a name
+func (b *WorkloadIdentityService) GetWorkloadIdentity(
+	ctx context.Context, name string,
+) (*workloadidentityv1pb.WorkloadIdentity, error) {
+	resource, err := b.service.GetResource(ctx, name)
+	return resource, trace.Wrap(err)
+}
+
+// ListWorkloadIdentities lists all WorkloadIdentitys using a given page size
+// and last key.
+func (b *WorkloadIdentityService) ListWorkloadIdentities(
+	ctx context.Context, pageSize int, currentToken string,
+) ([]*workloadidentityv1pb.WorkloadIdentity, string, error) {
+	r, nextToken, err := b.service.ListResources(ctx, pageSize, currentToken)
+	return r, nextToken, trace.Wrap(err)
+}
+
+// DeleteWorkloadIdentity deletes a specific WorkloadIdentitys.
+func (b *WorkloadIdentityService) DeleteWorkloadIdentity(
+	ctx context.Context, name string,
+) error {
+	return trace.Wrap(b.service.DeleteResource(ctx, name))
+}
+
+// DeleteAllWorkloadIdentities deletes all SPIFFE resources, this is typically
+// only meant to be used by the cache.
+func (b *WorkloadIdentityService) DeleteAllWorkloadIdentities(
+	ctx context.Context,
+) error {
+	return trace.Wrap(b.service.DeleteAllResources(ctx))
+}
+
+// UpsertWorkloadIdentity upserts a WorkloadIdentitys. Prefer using
+// CreateWorkloadIdentity. This is only designed for usage by the cache.
+func (b *WorkloadIdentityService) UpsertWorkloadIdentity(
+	ctx context.Context, resource *workloadidentityv1pb.WorkloadIdentity,
+) (*workloadidentityv1pb.WorkloadIdentity, error) {
+	upserted, err := b.service.UpsertResource(ctx, resource)
+	return upserted, trace.Wrap(err)
+}
+
+// UpdateWorkloadIdentity updates a specific WorkloadIdentity.
```

Comment:
Should we mention that it's a conditional update?

Reply:
9fb2f88611